cosigner/authcosigner_test.go

```diff
@@ -0,0 +1,252 @@
+package cosigner
+
+import (
+	"crypto"
+	"fmt"
+	"sync/atomic"
+	"testing"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/openpubkey/openpubkey/client"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/pktoken/mocks"
+	"github.com/openpubkey/openpubkey/util"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAuthIDs(t *testing.T) {
+	cosAlg := jwa.ES256
+	cosSigner, err := util.GenKeyPair(cosAlg)
+	if err != nil {
+		t.Error(err)
+	}
+	hmacKey := []byte{0x1, 0x2, 0x3}
+
+	cos := AuthCosigner{
+		Cosigner: Cosigner{
+			alg:    cosAlg,
+			signer: cosSigner,
+		},
+		Issuer:       "https://example.com",
+		KeyID:        "kid1234",
+		authIdIter:   atomic.Uint64{},
+		hmacKey:      hmacKey,
+		AuthStateMap: make(map[string]*AuthState),
+		AuthCodeMap:  make(map[string]string),
+	}
+
+	// Test if we get the same value if we supply exact the same time
+	unixTime := uint64(5)
+	authID1, err := cos.CreateAuthID(unixTime)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	authID2, err := cos.CreateAuthID(unixTime)
+	if err != nil {
+		t.Fatal(err)
+	}
+	require.NotEqualValues(t, authID1, authID2)
+
+	require.Equal(t, "644117927902f52d3949804c7ce417509d9437eb1240a9bf75725c9f61d5b424", authID1)
+	require.Equal(t, "f7d16adcef9f7d0e72139f0edae98db64c2db1f0cb8b59468d4766e91126f4eb", authID2)
+}
+
+func TestInitAuth(t *testing.T) {
+	cos := CreateAuthCosigner(t)
+
+	alg := jwa.ES256
+	signer, err := util.GenKeyPair(alg)
+	if err != nil {
+		t.Error(err)
+	}
+	pkt, err := mocks.GenerateMockPKToken(signer, alg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cosP := client.CosignerProvider{
+		Issuer:       "example.com",
+		CallbackPath: "/mfaredirect",
+	}
+	redirectURI := fmt.Sprintf("%s/%s", "http://localhost:5555", cosP.CallbackPath)
+
+	initAuthMsgJson, _, err := cosP.CreateInitAuthSig(redirectURI)
+	sig, err := pkt.NewSignedMessage(initAuthMsgJson, signer)
+	authID1, err := cos.InitAuth(pkt, sig)
+	if err != nil {
+		t.Error(err)
+	}
+	require.NotEmpty(t, authID1)
```

Comment:
Based on the changes detailed below I am marking this as fixed. Open a new comment if there are still changes you want here.

> I only ask because it makes this unit test more isolated to the cosigner package if we don't have to use the client package in order to test AuthCosigner.

This test is doing double duty right now, testing both AuthCosigner and CosignerProvider message construction. Longer term I'd like to fully isolate them, but I don't think it is worth it until we have Cosign functionality locked down.

CallbackPath is used as part of the redirectURI sent the the Cosigner. Based on your comment I have added the following check to CreateInitAuthSig

```golang
	if !strings.HasSuffix(redirectURI, c.CallbackPath) {
		return nil, "", fmt.Errorf("redirectURI (%s) does not end in expected callbackPath (%s)", redirectURI, c.CallbackPath)
	}
```

Regarding the issuer issue: thinking about it more, I've come to the conclusion that the issuer should be in the signed message and I've made that change.

```golang
type InitMFAAuth struct {
	Issuer      string `json:"iss"`
	RedirectUri string `json:"ruri"`
	TimeSigned  int64  `json:"time"`
	Nonce       string `json:"nonce"`
}
```

This rules out cross-issuer replay attacks since the cosigner will discard a signed message that doesn't include the issuer. Your comment was very helpful because once after adding this check on the issuer a bunch of tests failed where I had the wrong issuer. It is a good sanity check to have and improves security.